scripts/binaryninja/manticore_viz/__init__.py

```diff
@@ -0,0 +1,213 @@
+import fcntl
+import os
+import time
+import threading
+
+from binaryninja import (
+    PluginCommand, HighlightStandardColor, log, BackgroundTaskThread
+)
+from binaryninja.interaction import (
+    get_open_filename_input, get_directory_name_input, get_choice_input
+)
+import binaryninja.enums as enums
+
+blue = HighlightStandardColor.BlueHighlightColor
+black = HighlightStandardColor.BlackHighlightColor
+white = HighlightStandardColor.WhiteHighlightColor
+clear = HighlightStandardColor.NoHighlightColor
+# renew interval
+interval = 3
+
+class Singleton(type):
+    _instances = {}
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+class TraceVisualizer(object):
+    __metaclass__ = Singleton
+    def __init__(self, view, workspace, base=0x0, live=False):
+        self.view = view
+        self.workspace = workspace
+        self.base = base
+        # highlighted addresses
+        self.highlighted = set()
+        # covered basic blocks
+        self.cov_bb = set()
+        # comments inserted
+        self.cov_comments = set()
+        self.current_function = None
+        self.live_update = live
+
+    def visualize(self):
+        """
+        Given a Manticore workspace, or trace file, highlight the basic blocks.
+        """
+        if os.path.isfile(self.workspace):
+            t = threading.Thread(target=self.highlight_from_file,
+                                 args=(self.workspace,))
+        elif os.path.isdir(self.workspace):
+            t = threading.Thread(target=self.highlight_from_dir,
+                                 args=(self.workspace,))
+        t.start()
+
+    def highlight_from_file(self, tracefile):
+        while True:
+            self.process_trace(tracefile)
+            if not self.live_update:
+                break
+            time.sleep(interval)
+
+    def highlight_from_dir(self, workspace_dir):
+        while True:
```

Comment:
since it looks like this live update thread will infinite loop thread, can we have a way to let the user end live visualization mode? maybe there could be "start live update" and "end live update" buttons